File: rest-common/src/main/java/org/chronopolis/rest/models/storage/StagingCreate.java
package org.chronopolis.rest.models.storage;

import org.chronopolis.rest.support.StorageUnit;
import org.hibernate.validator.constraints.NotBlank;

import javax.validation.constraints.Min;
import javax.validation.constraints.NotNull;

/**
 * Information for form data which is used to create a StagingStorage model/entity/whatever
 *
 * @author shake
 */
public class StagingCreate {

    /**
     * The relative location of the data which has been staged
     */
    @NotBlank
    private String location;

    /**
     * The ID of the StorageRegion which data is staged in
     */
    @NotNull
    private Long storageRegion;

    /**
     * The number of files staged
     */
    @Min(1L)
    @NotNull
    private Long totalFiles;

    /**
     * The size on disk of the data
     */
    @Min(1L)
    @NotNull
    private Long size;

    /**
     * The StorageUnit for the size
     */
    @NotNull
    private StorageUnit storageUnit;

    public String getLocation() {
        return location;
    }

    public StagingCreate setLocation(String location) {
        this.location = location;
        return this;
    }

    public Long getStorageRegion() {
        return storageRegion;
    }

    public StagingCreate setStorageRegion(Long storageRegion) {
        this.storageRegion = storageRegion;
        return this;
    }

    public Long getTotalFiles() {
        return totalFiles;
    }

    public StagingCreate setTotalFiles(Long totalFiles) {
        this.totalFiles = totalFiles;
        return this;
    }

    public Long getSize() {
        return size;
    }

    public StagingCreate setSize(Long size) {
        this.size = size;
        return this;
    }

    public StorageUnit getStorageUnit() {
        return storageUnit;
    }

    public StagingCreate setStorageUnit(StorageUnit storageUnit) {
        this.storageUnit = storageUnit;
        return this;
    }
}

File: rest-models/src/main/kotlin/org/chronopolis/rest/models/create/StagingCreate.kt
package org.chronopolis.rest.models.create

import org.chronopolis.rest.models.Bag
import org.chronopolis.rest.models.StagingStorage
import org.chronopolis.rest.models.StorageRegion
import org.chronopolis.rest.models.enums.StorageUnit
import org.hibernate.validator.constraints.NotBlank
import javax.validation.constraints.Min

/**
 * Request to create [StagingStorage] for a [Bag]
 *
 * @property location the relative location of the staged data within the [StorageRegion]
 * @property validationFile the name of the [DataFile] used to validate transfers
 * @property storageRegion the [StorageRegion] the staged data resides in
 * @property totalFiles the number of files staged
 * @property size the amount of storage capacity used by the staged data
 * @property storageUnit the [StorageUnit] used in conjunction with the [size] to define the usage
 */
data class StagingCreate(@get:NotBlank var location: String = "",
                         @get:NotBlank var validationFile: String = "/tagmanifest-sha256.txt",
                         @get:Min(1) var storageRegion: Long = -1,
                         @get:Min(1) var totalFiles: Long = -1,
                         @get:Min(1) var size: Long = -1,
                         var storageUnit: StorageUnit = StorageUnit.B)